src/server.rs

```diff
@@ -563,34 +578,261 @@ impl ServerState {
         return true;
     }
 
+    // Request_ack contains the last processed sequence number. If one arrives older (less than)
+    // than the last processed, it must be rejected.
+    // FIXME Does not handle wrapped sequence number case yet.
+    fn is_previously_processed_packet(&mut self, player_id: PlayerID, sequence: u64) -> bool {
+        let player: &Player = self.get_player(player_id);
+        if let Some(request_ack) = player.request_ack {
+            if sequence <= request_ack {
+                return true;
+            }
+        }
+        false
+    }
+
     fn get_player_id_by_cookie(&self, cookie: &str) -> Option<PlayerID> {
         match self.player_map.get(cookie) {
             Some(player_id) => Some(*player_id),
             None => None
         }
     }
 
-    // always returns either Ok(Some(Packet::Response{...})), Ok(None), or error
+    /// Returns true if the packet already exists in the queue, otherwise it will return false, and
+    /// will be added in sequence_number order.
+    fn add_packet_to_queue(&mut self, player_id: PlayerID, packet: Packet) -> bool {
+        // Unwrap should be safe since a player ID was already found.
+        let network: &mut NetworkManager = self.network_map.get_mut(&player_id).unwrap();
+        let already_exists = network.rx_packets.buffer_item(packet);
+        already_exists
+    }
+
+    /// Checks to see if the incoming packet is immediately processable
+    fn can_process_packet(&mut self, player_id: PlayerID, sequence_number: u64) -> bool {
+        let player: &mut Player = self.get_player_mut(player_id);
+        if let Some(ack) = player.request_ack {
+            trace!("[CAN PROCESS?] Ack: {} Sqn: {}", ack, sequence_number);
+            ack+1 == sequence_number
+        } else {
+            // request_ack has not been set yet, likely first packet
+            player.request_ack = Some(0);
+            true
+        }
+    }
+
+    /// Processes a player's request action for all non Connect requests. If necessary, a response is buffered
+    /// for later transmission
+    fn process_player_request_action(&mut self, player_id: PlayerID, action: RequestAction) -> Result<Option<Packet>, Box<Error>> {
```

Comment:
Maybe it's just me, but I got a little confused by the long chain of function calls, each handling only one type of message.
process_player_request_action -> handle_request_action -> process_request_action

Reply:
Definitely not just you. Will look into clarifying that call-chain.